src/debug_utils.cc

```diff
@@ -299,6 +299,106 @@ void CheckedUvLoopClose(uv_loop_t* loop) {
   CHECK(0 && "uv_loop_close() while having open handles");
 }
 
+std::vector<std::string> NativeSymbolDebuggingContext::GetLoadedLibraries() {
+  std::vector<std::string> list(0);
+#ifdef __linux__
+  dl_iterate_phdr(
+      [](struct dl_phdr_info* info, size_t size, void* data) {
+        std::vector<std::string>* list = static_cast<std::vector<std::string>*>(data);
+        if (info->dlpi_name != nullptr && *info->dlpi_name != '\0') {
+          list->push_back(info->dlpi_name);
+        }
+        return 0;
+      },
+      &list);
+#elif __APPLE__
+  int i = 0;
+  for (const char* name = _dyld_get_image_name(i); name != nullptr;
+       name = _dyld_get_image_name(++i)) {
+    list.push_back(name);
+  }
+
+#elif _AIX
+  // We can't tell in advance how large the buffer needs to be.
+  // Retry until we reach too large a size (1Mb).
+  const unsigned int buffer_inc = 4096; 
+  unsigned int buffer_size = buffer_inc;
+  MallocedBuffer<char> buffer(buffer_size);
+  int rc = -1;
+  while (rc != 0 && buffer_size < 1024 * 1024) { 
+    rc = loadquery(L_GETINFO, buffer.data, buffer_size);
+    if (rc == 0) { 
+      break;  
+    }
+    buffer_size += buffer_inc;
+    buffer = MallocedBuffer<char>(buffer_size);
+  }
+  if (rc == 0) { 
+    char* buf = buffer.data;
+    ld_info* cur_info = nullptr;
+    do {    
+      std::ostringstream str;
+      cur_info = reinterpret_cast<ld_info*>(buf);
+      char* member_name = cur_info->ldinfo_filename +
+          strlen(cur_info->ldinfo_filename) + 1;
+      if (*member_name != '\0') { 
+        str << cur_info->ldinfo_filename << "(" << member_name << ")"; 
+        list.push_back(str.str());
+        str.str("");
+      } else {
+        list.push_back(cur_info->ldinfo_filename);
+      }
+      buf += cur_info->ldinfo_next;
+    } while (cur_info->ldinfo_next != 0);
+  }
+#elif __sun
+  Link_map* p;
+
+  if (dlinfo(RTLD_SELF, RTLD_DI_LINKMAP, &p) != -1) {
+    for (Link_map* l = p; l != nullptr; l = l->l_next) {
+      list.push_back(l->l_name);
+    }
+  }
+
+#elif _WIN32
+  // Windows implementation - get a handle to the process.
+  HANDLE process_handle = OpenProcess(PROCESS_QUERY_INFORMATION|PROCESS_VM_READ,
+                                      FALSE, GetCurrentProcessId());
+  if (process_handle == nullptr) {
+    // Cannot proceed, return an empty list.
+    return list;
+  }
+  // Get a list of all the modules in this process
+  DWORD size_1 = 0, size_2 = 0;
+  // First call to get the size of module array needed
+  if (EnumProcessModules(process_handle, nullptr, 0, &size_1)) {
+    MallocedBuffer<HMODULE> modules(size_1);
+
+    // Second call to populate the module array
+    if (EnumProcessModules(process_handle, modules.data, size_1, &size_2)) {
+      for (int i = 0;
+           i < (size_1 / sizeof(HMODULE)) && i < (size_2 / sizeof(HMODULE));
```

Comment:
Do you know if `size_1 != size_2` is a real possibility?

Reply:
I don't know; my reasoning is: if modules get loaded (or unloaded) by the OS asynchronously with the process, there could be tiny race window between the first and second query.

If you think that is an extreme step, or we could be more conservative, then we could iterate over a count that is smaller of the two; what do you say?

Reply:
No, this sounds fine to me :+1: